[SYCL][Doc] Update C++ standard support in GSG (#3478)
intel/llvm: sycl/doc/GetStartedGuide.md

## C++ standard

* DPC++ headers require C++17 at least.

> * DPC++ runtime requires C++14 at least.

Reviewer: According to my understanding, DPC++ runtime uses DPC headers, which means if C++ version used in DPC++ header is higher than C++ used to build DPC++ runtime, it might not compile.

In order words DPC++ headers C++ version must be <= DPC++ runtime C++ version.

Reviewer 2: I agree. If we want uplift we should uplift both having remaining issues at #3447 resolved.

Author: С++17 code in headers is not used in the library and is guarded by macro `#if __cplusplus > 201402L`, so the current combination of headers (which have some C++17 code) and library (C++14) works.

Reviewer 2: I think you are correct, however it's sort of library internal knowledge, raising both looks more consistent.

Reviewer: > С++17 code in headers is not used in the library and is guarded by macro `#if __cplusplus > 201402L`, so the current combination of headers (which have some C++17 code) and library (C++14) works.

That means that headers require C++14, not C++17.

Author: @bader, do you mean that adding C++17 code to the headers is some kind of hack now, and in this case the headers officially still require the same C++14 standard?

Reviewer: > do you mean that adding C++17 code to the headers is some kind of hack now

No at all. Making use of C++17 features in DPC++ headers is not a hack.

> in this case the headers officially still require the same C++14 standard

DPC++ runtime is compiled in C++14 mode, which headers headers must be compatible with C++14 mode.

Author: Ok, I see. In this case I can leave this information w/o any changes and it should be updated in https://github.com/intel/llvm/pull/3447

> * DPC++ compiler supports building apps as C++17 apps and higher.
> * DPC++ compiler builds apps as C++17 apps by default.

Reviewer 2: ```suggestion
* DPC++ compiler builds apps as C++17 apps by default. Higher versions of standard are supported as well.
```

Author: Fixed
